crates/oxc_traverse/src/context/scoping.rs

```diff
@@ -121,7 +121,7 @@ impl TraverseScoping {
     /// `flags` provided are amended to inherit from parent scope's flags.
     pub fn create_scope_child_of_current(&mut self, flags: ScopeFlags) -> ScopeId {
         let flags = self.scopes.get_new_scope_flags(flags, self.current_scope_id);
-        self.scopes.add_scope(Some(self.current_scope_id), flags)
+        self.scopes.add_scope(Some(self.current_scope_id), AstNodeId::dummy(), flags)
```

Comment:
I think this will result in multiple scopes having the same node ID (0), is this a bug in the logic?

Reply:
I can answer this one. `Traverse` is a mutating visitor. For this reason, it does not use all of `Semantic`. It can't use `AstNodes` because that contains `AstKind`s which contain references to all actual AST nodes. Holding that would be a violation of Rust's aliasing rules, because visitor functions also get passed `&mut` references to AST nodes => would result in having a live `&` and `&mut` reference to same AST node simultaneously.

So... long story short... `Traverse` uses only an incomplete subset of `Semantic` and that doesn't include `AstNodes`, so it uses dummy `AstNodeId`s.

This is part of what I hope to look at when I review `Semantic`. We need some way to merge the 2 paradigms.